### ISSUE TITLE
False alert on needing counter cache

### ISSUE THREAD

**Issue by iczman** [flyerhzm/bullet#288]
Seems to be a regression on 5.1.0, as it doesn't reproduce on 5.0.0

**Product Model**

```
class Product < ActiveRecord::Base
  has_many :images, class_name: "ProductImage", dependent: :destroy
end
```

**ProductImage Model**

```
class ProductImage < ActiveRecord::Base
  belongs_to :product
  mount_uploader :image, ImageUploader
end
```

**Product Controller**

```
class Member::ProductsController < Member::ApplicationController
  def index
    @products = Product.all.includes(:images)
  end
end
```

**False Alert**

```
Need Counter Cache
  Product => [:images]
```

**Rails Log**

```
Started GET "/member/admin/products" for ::1 at 2016-05-27 16:30:28 +0800
Processing by Member::Admin::ProductsController#index as HTML
  Member::User Load (0.3ms)  SELECT  `member_users`.* FROM `member_users` WHERE `member_users`.`id` = 1  ORDER BY `member_users`.`id` ASC LIMIT 1
  Rendered member/admin/products/_list_group.html.erb (0.2ms)
  Product Load (0.3ms)  SELECT `products`.* FROM `products`
  ProductImage Load (0.2ms)  SELECT `product_images`.* FROM `product_images` WHERE `product_images`.`product_id` IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  Rendered member/admin/products/index.html.erb within layouts/member_admin (58.8ms)
  Rendered common/member/admin/_header.html.erb (1.1ms)
  Rendered common/member/admin/_footer.html.erb (0.1ms)
Completed 200 OK in 302ms (Views: 299.0ms | ActiveRecord: 0.9ms)
  Rendered /usr/local/lib/ruby/gems/2.3.0/gems/xray-rails-0.1.21/app/views/_xray_bar.html.erb (10.4ms)
```

Note: I don't use anything like Product.images.size anywhere in the codebase that would kick off a SQL COUNT(*).

**Comment by flyerhzm** [comment]
@iczman do you call `empty?` on images?

**Comment by iczman** [comment]
Ah I do have images.any? in my Product model. Does this meet the requirement for needing counter cache?

**Comment by flyerhzm** [comment]
@iczman no, you don't need a counter cache if it doesn't generate a COUNT sql, it is a false alert, I'll fix

**Comment by iczman** [comment]
@flyerhzm thanks for the quick reply.

**Comment by oniofchaos** [comment]
@flyerhzm was this ever fixed? I'm on 5.4.2 and experiencing the same issue, locking at 5.0 fixed it.

**Comment by flyerhzm** [comment]
I added the following test code to spec/integration/counter_cache_spec.rb

```
it "should not need counter cache without size" do
  Country.includes(:cities).each do |country|
    country.cities.any? { |city| city.name }
  end
  expect(Bullet.collected_counter_cache_notifications).to be_empty
end
```

and it passed, could you let me know how it happened?

**Comment by oniofchaos** [comment]
Looks like it's being triggered by an `empty?`

`model.association.empty?`

**Comment by flyerhzm** [comment]
@oniofchaos I tried to use `empty?` instead of `any?`, it's also passed, can't reproduce it.

**Comment by oniofchaos** [comment]
What can I do to help? The code is a private repo so I can't share directly. When I remove the `empty?` check, the Bullet error goes away. When I go back to 5.0 I think it was, the error also goes away.

**Comment by flyerhzm** [comment]
@oniofchaos it would be better if you can create a demo app and reproduce such issue for me.

**Comment by oniofchaos** [comment]
In a new project, I was unable to reproduce the issue sadly.

Any tips for what could be causing this? One thing I realized today is that the association is polymorphic. Not sure if that makes a difference.

**Comment by meganemura** [comment]
@flyerhzm 
I faced same error with rails 4.2 and bullet 5.4.3.

I added [test code](https://github.com/flyerhzm/bullet/issues/288#issuecomment-255587133) and replaced `any? {...}` with `empty?`.

With Rails 5.0, rspec did not report any errors. But with older Rails (<= 4.2), rspec reported below.
```
$ BUNDLE_GEMFILE=Gemfile.rails-4.2 bundle exec rspec
(snip)
Failures:

  1) Bullet::Detector::CounterCache should not need counter cache without size
     Failure/Error: expect(Bullet.collected_counter_cache_notifications).to be_empty
       expected `[#<Bullet::Notification::CounterCache:0x007fc9c3ac14a0 @base_class="Country", @associations=[:cities], @path=nil>].empty?` to return true, got false
     # ./spec/integration/counter_cache_spec.rb:37:in `block (2 levels) in <top (required)>'
```

modified test code in `spec/integration/counter_cache_spec.rb`:
```ruby
it "should not need counter cache without size" do
  Country.includes(:cities).each do |country|
    country.cities.empty?
  end
  expect(Bullet.collected_counter_cache_notifications).to be_empty
end
```

**Comment by flyerhzm** [comment]
@oniofchaos @meganemura I have fixed it in bullet 5.5.0
